File: 0707-design-linked-list/0707-design-linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class MyLinkedList:

    def __init__(self):
        self.head = self.tail = None
        self.length = 0
        
    def traverse(self, index: int) -> Node:
        position = 0
        current_node = self.head
        while position != index:
            current_node = current_node.next
            position += 1
        return current_node

    def get(self, index: int) -> int:
        
        if index < 0 or index >= self.length:
            return -1
        current_node = self.traverse(index)
        return current_node.value

    def addAtHead(self, value: int) -> None:
        new_node = Node(value)
        if not self.head:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.length += 1

    def addAtTail(self, value: int) -> None:
        new_node = Node(value)
        if not self.head:
            self.head = self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1

    def addAtIndex(self, index: int, value: int) -> None:
        if index < 0 or index > self.length:
            return
        if index == 0:
            self.addAtHead(value)
            return
        if index == self.length:
            self.addAtTail(value)
            return
        new_node = Node(value)
        previous_node = self.traverse(index - 1)
        new_node.next = previous_node.next
        previous_node.next = new_node
        self.length += 1


    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.length:
            return
        if index == 0:
            self.head = self.head.next
            if self.length == 1:
                self.tail = None
        else:
            previous_node = self.traverse(index - 1)
            current_node = previous_node.next
            previous_node.next = current_node.next
            if index == self.length - 1:
                self.tail = previous_node
        self.length -= 1

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 0707-design-linked-list/0707-design-linked-list.java
class Node{
    int value;
    Node next;

    Node(int value){
        this.value = value;
        this.next = null;
    }
}

class MyLinkedList {
    Node head;
    Node tail;
    int length;

    public MyLinkedList() {
        this.head = this.tail = null;
        this.length = 0;
    }
    
    public Node traverse(int index){
        int position = 0;
        Node currentNode = this.head;
        while(position != index){
            currentNode = currentNode.next;
            position++;
        }
        return currentNode;
    }
    
    public int get(int index) {
        if(index < 0 || index >= this.length){
            return -1;
        }
        Node currentNode = traverse(index);
        return currentNode.value;
    }
    
    public void addAtHead(int value) {
        Node newNode = new Node(value);
        if(this.head == null){
            this.head = this.tail = newNode;
        }
        else{
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
    }
    
    public void addAtTail(int value) {
        Node newNode = new Node(value);
        if(this.head == null){
            this.head = this.tail = newNode;
        }
        else{
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }
    
    public void addAtIndex(int index, int value) {
        if(index < 0 || index > this.length) {
            return;
        }
        if(index == 0) {
            addAtHead(value);
            return;
        }
        if(index == length){
            addAtTail(value);
            return;
        }
        Node newNode = new Node(value);
        Node previousNode = traverse(index - 1);
        newNode.next = previousNode.next;
        previousNode.next = newNode;
        this.length++;
    }
    
    public void deleteAtIndex(int index) {
        if (this.head == null || index < 0 || index >= this.length) {
            return;
        }
        if (index == 0) {
            this.head = this.head.next;
            if (this.length == 1) {
                this.tail = null;
            }
        } else {
            Node previousNode = traverse(index - 1);
            Node currentNode = previousNode.next;
            previousNode.next = currentNode.next;
            if (index == this.length - 1) {
                this.tail = previousNode;
            }
        }
        this.length--;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */